### DISCUSSION TITLE
Chatting from within DOSEMU2

### DISCUSSION THREAD

**Discussion by jschwartzenberg** [dosemu2/dosemu2#2234] [Off-topic]
So this works:
![Screenshot_20240709_221350](https://github.com/dosemu2/dosemu2/assets/877112/b6781f02-4e1d-46a1-b3b3-92172e7deedc)

You can add me: jschwart

I'm curious to hear if there are other clients that can run inside DOSEMU2!

Edit: maybe some instructions are helpful.
1. Install Netscape 4.08 inside DOSEMU2
2. Sign up at https://phoenixim.ddns.net/ (probably that won't work through Netscape :P)
3. Open AIM and change the server from `login.oscar.aol.com` to `phoenixim.ddns.net`
4. Connect with the credentials you signed up with

**Comment by stsp** [comment]
These instructions are not enough.
You seem to use twinsock, which is
interesting. Trumpet is proprietary
IIRC?
I wonder if @javispedro have some
winsock solution for win31?
Otherwise it might be possible to
embed twinsock into dosemu2, but
that's a bit of work.

**Reply by jschwartzenberg** [reply]
Yes, I use Ultra TwinSock. (TwinSock 2.0 doesn't work for some reason.)
Instructions for that:
- Download Ultra TwinSock 0.2: [ultrts02.zip](https://github.com/user-attachments/files/16150720/ultrts02.zip)
- Unpack
- Run `./build.sh` to build the `tshost` binary on the host side
- Put `$_com4 = "exec bash"` in `~/.dosemurc`
- Run DOSEMU2 with `-no-priv-sep`
- Copy the bundled `winsock.dll` to `c:\windows\`
- Run Ultra TwinSock and set the communications as follows:
![Screenshot_20240709_232531](https://github.com/dosemu2/dosemu2/assets/877112/25b88ceb-211c-411f-909a-1052586b3277)
- Then (blindly) type whatever commands needs in the Ultra TwinSock window (which will be your Bash session at this point) to start `tshost`
- Ultra TwinSock will automatically minimize and you have a connection

**Reply by stsp** [reply]
It has weird license.
It might be better to use twinsock.
Do they have incompatible protocols?
I.e. can thost be replaced between them?

**Reply by jschwartzenberg** [reply]
TwinSock 1.2 is GPLv2+, while later versions have a different license. It seems that I did not handle this properly when publishing sources on GitHub. If you want I can ask the author if he's willing to release all versions as GPLv2+. I had mail contact with him when I published the sources.

Here are the original zips of the different versions. It seems they became harder to find:
[twnsck12.zip](https://github.com/user-attachments/files/16168193/twnsck12.zip)
[TWNSCK13.ZIP](https://github.com/user-attachments/files/16168197/TWNSCK13.ZIP)
[twnsck14.zip](https://github.com/user-attachments/files/16168194/twnsck14.zip)
[twnsck20.zip](https://github.com/user-attachments/files/16168195/twnsck20.zip)

See also: https://web.archive.org/web/20190423161627/http://www.troy.rollo.name/opensource.html

**Reply by jschwartzenberg** [reply]
Regarding the protocols, it seems there were some changes. At least the 2.0 documentation documents they were changed.

**Reply by stsp** [reply]
I am not sure if we need any
particular version under gpl.
We certainly need some version
that builds. If we embed it into
dosemu2, the transport will have
to be rewritten anyway, which is
why it would just be a starting point,
so whatever version builds - suits.

**Reply by jschwartzenberg** [reply]
You mean just building the host side? (Not the client/Windows side?)
At least the host part of Ultra TwinSock and TwinSock 2.0 builds without changes on modern GNU/Linux.

**Reply by stsp** [reply]
No, I mean windows side.
The host side will likely be rewritten.
Ideally we should use trumpet's spec
for transport, but with windows side
replaced with GPLed one.

**Reply by jschwartzenberg** [reply]
What does trumpet's spec for transport refer to? Doesn't Trumpet WinSock rely on PPP/SLIP or a packet driver to communicate with the outside world?

With regarding building TwinSock on Windows, I guess we should see which win31 compilers we can run (on DOSEMU2).

I corrected the licenses in my repo now. Version 1.0 and 1.2 are GPLv2+ in any case.

**Reply by stsp** [reply]
The spec is here:
https://github.com/dosemu2/dosemu2/issues/167#issuecomment-1744603994
I don't know the exact impl of
the trumpet winsock. Maybe it
consists of 2 parts: 1 implements
this spec on top of the packet driver,
and another one is winsock itself.
Or maybe this spec is completely
omitted in trumpet winsock and is
only used in trumpet DOS products.

**Reply by jschwartzenberg** [reply]
It says this:
> Preliminary Specification for a TCP/IP Application Binary Interface (ABI) to run over MS-DOS.

I think this would be the interface between DOS applications and their DOS library.

**Reply by stsp** [reply]
Yes, but if we add that, we can as well
use it with the custom winsock. This
spec is the most simple of all available.

**Reply by jschwartzenberg** [reply]
So DOSEMU2 would expose something compatible with the Trumpet DOS spec? And then WinSock would use the same spec being exposed? Did you look into implementing the first part already? I guess it would be a plugin?
We mainly need a compiler that can compile 16-bit NE DLLs for the WinSock part?

**Reply by stsp** [reply]
> So DOSEMU2 would expose something compatible with the Trumpet DOS spec?
> And then WinSock would use the same spec being exposed?

We will need to code up such winsock
from the GPLed ones.

> Did you look into implementing the first part already?

Not yet.
Too many other things not implemented. :)

> I guess it would be a plugin?

We usually make plugins to the external
libraries. Even fdpp and dj64dev are the
external libraries.
Interrupt handlers are not plugins. It can
be structured similar to packet driver.

> We mainly need a compiler that can compile 16-bit NE DLLs for the WinSock part?

Whatever compiles twinsock, should suite.
We already compiled the mouse driver some
time in the past.

**Reply by stsp** [reply]
> Did you look into implementing the first part already?

In fact, yes, I looked. :)
https://ftp.sunet.se/mirror/archive/ftp.sunet.se/pub/simtelnet/trumpet/tcp-abi/
Here you can find the collection of
DOS trumpet apps, and the interesting
part is ntcpdrv.zip. This is exactly the
"first part" you mentioned.

**Reply by stsp** [reply]
Oh, and twinsock is not GPL, either.
So any GPLes winsock there actually?

**Reply by jschwartzenberg** [reply]
TwinSock up to 1.2 is GPLv2+. Wine's version might be better/easier though :)

**Reply by stsp** [reply]
> TwinSock up to 1.2 is GPLv2+.

Is it in your repo?

**Reply by jschwartzenberg** [reply]
> Is it in your repo?

Yes:
https://github.com/jschwartzenberg/twinsock/tree/v1.2
https://github.com/jschwartzenberg/twinsock/releases/tag/v1.2

I collected and pushed all archives of TwinSock I could find and created tags for each version. The releases just have the original ZIPs attached. (I also took the original author name and e-mail address and timestamps from the archives to provide an author and date for each commit ;)

**Reply by stsp** [reply]
1.4 also seems GPL.
Well, license change deserves a new
branch, not a tag.
Would be good if you put a non-GPL
thing to a separate branch.

**Reply by jschwartzenberg** [reply]
It seems I didn't push the new tags properly when I fixed the licenses for the newer versions...
Here is the license for 1.3: https://github.com/jschwartzenberg/twinsock/commit/3f52e86b3330a425691dde3b362249f60832c228

I'll look into fixing the tags.

Separate branches for the other licenses seems like a good idea. I might just ask the author if he could declare all releases to be GPLv2+ though.

**Reply by stsp** [reply]
Or even GPLv3+ :)
Although I am not sure his
university e-mail address is
still valid...

**Reply by jschwartzenberg** [reply]
That address is not valid anymore, but back in 2017 it was not hard to find his current e-mail address and reach him :)

**Comment by stsp** [comment]
Hmm, actually the perfectly GPLed
winsock is here:
https://github.com/wine-mirror/wine/blob/master/dlls/winsock.dll16/socket.c

@stefand how difficult would it be
to convert that source to the actual
winsock.dll? I realize that the transport
will have to be added, as currently it
uses the host's socket API. But this is
a minor problem. I think the bigger
problem is to build it under something
like OpenWatcom?
Any suggestions?

**Reply by stsp** [reply]
Also an OT: have you received my
suggestion (by e-mail) how to duplicate
mmap()s with mremap()? Haven't heard
from you since. :)

**Reply by stefand** [reply]
I saw your E-Mail but guess was distracted by something else. I'll keep your suggestions in mind the next time the discussion pops up on the Wine lists :-)

**Reply by stefand** [reply]
I don't think wine's Winsock will help you much. As you noticed it is built pretty much around passing it to Unix socket functions. Beyond that nowadays our winsock / ws2_32 DLLs are first calling into an ntdll / ntoskrnl layer that then does the Unix calls. I don't know how good our support for the win16 winsock is - in theory it shouldn't be too different from Win32, but we have zero 16 bit specific test coverage.

**Reply by stsp** [reply]
> As you noticed it is built pretty much around passing it to Unix socket functions.

Yes, but the question was mostly
"how difficult is it to turn that into an actual
winsock.dll", as for me this is the hardest
part (since I have very few experience
compiling things for windows). The code
is not big, so replacing native socket calls
with some proxy layer is not a problem at
all. Building is the problem.
But I'll try to avoid that if we have an
alternative impl.

**Reply by stsp** [reply]
Knowing that wine now itself uses mingw
AFAIK to compile things as PEs, I suppose
the task of getting a winsock.dll is probably
not too hard? Basically just unbinding it
from wine internals, I suppose?

**Reply by stefand** [reply]
ws2_32 is already a PE DLL.  It talks to the "kernel" via ioctls that go to afd.sys. You'd have to strip the AFD ioctls out and replace it with something that either uses the DOS packet driver API (thus re-implement Trumpet Winsock), proxies it to the Linux side (thus re-implement twinsock) or build an entire NDIS API layer.

I don't know how much is left of Wine's winsock after you remove those ioctls. Probably just the function definitions. I don't think it is worth it.

The other trouble you'll face is that Wine doesn't compile to any 16 bit code. Even our foo.dll16 libs are 32 bit, just with some auto-generated 16<->32 bit thunks. That won't work in Windows 3.x, at least not without Win32s.

My 2c is if you want to be independent of Trumpet Winsock, look into Twinsock. It sounds closest to what you want.

**Reply by stsp** [reply]
Yep, the problem with twinsock
was that its not GPL. But as @jschwartzenberg 
explained, some versions are, plus
he says it might be possible to get
to its author. So since wine's doesn't
suite, I'll use that one.

> it with something that either uses the DOS packet driver API (thus re-implement Trumpet Winsock),
> proxies it to the Linux side (thus re-implement twinsock) or build an entire NDIS API layer.

My idea was to use ntcpdrv.zip from here:
https://ftp.sunet.se/mirror/archive/ftp.sunet.se/pub/simtelnet/trumpet/tcp-abi/
which is also very well documented:
https://ftp.sunet.se/mirror/archive/ftp.sunet.se/pub/simtelnet/trumpet/tcp-abi/tcp201.spe
Eventually this spec can be embedded,
and ntcpdrv.zip can then go away.
It would be cool to have NDIS api too,
for running msclient (smb client), but
unfortunately I have zero idea how NDIS2
works or where to steal the code from.
If you know, please speak up. :)

**Comment by stsp** [comment]
So I added the trumpet API now.
Someone may try to port twinsock
on that. :)

**Comment by stefand** [comment]
So how does one use this trumpet API? Does trumpet winsock have some way to talk to it instead of the packet driver API? Are there DOS applications that use it?

From looking at it my guess is that whatever application is using this API can directly listen() and connect() from the host API. In other words, there's no need to set up a TUN device with its own IP and set up IP forwarding, masquerading and port forwarding. Do I understand this correctly?

**Reply by stsp** [reply]
> Does trumpet winsock have some way to talk to it instead of the packet driver API?

Haven't tried, but AFAIK this is not
the case. We need to code up the
replacement from twinsock.

> Are there DOS applications that use it?

Please see:
https://ftp.sunet.se/mirror/archive/ftp.sunet.se/pub/simtelnet/trumpet/tcp-abi/
https://github.com/dosemu2/dosemu2/pull/2253#issuecomment-2273252562
But this is not the point.
The point is to find some light-weight
tcp/ip suit, write the tcpabi back-end
for it, add it to dj64dev and build a few
open-source apps. Then there will be
no shortage of an apps. But, as usual,
this is a big task... Maybe you can join. :)

> Do I understand this correctly?

Yes, this is correct. No preparations needed.
And with `emutcp` tool you can control
what network interface to use.

**Comment by stsp** [comment]
Well, actually what we need for
winsock and the rest, is this:
https://www.richdawe.be/archive/dl/csockvxd.htm
Unfortunately it seems the sources
of the marvel are lost. I only managed
to figure out that it was the part of the
CodaFS client for win95, written by
Michael Callahan. The file with sources
was named "vc95.zip", but it seems to
be lost forever.

Anyway, even the available documentation
is a great finding. It might be possible
to reconstruct an API. And this is much,
much more plausible for winsock or dj64dev
than the tcpabi thingy. I think tcpabi should
be left alone with a couple of apps it has,
and get no further development.

**Reply by jschwartzenberg** [reply]
In the big dump that people are talking about these days, I found another stack that seems to have both DOS and Windows parts. No idea whether it might be interesting or relevant, but here is a copy:

[3000DOS1.ZIP](https://github.com/user-attachments/files/16567289/3000DOS1.ZIP)
[3000DOS2.ZIP](https://github.com/user-attachments/files/16567291/3000DOS2.ZIP)
[3000WN-1.ZIP](https://github.com/user-attachments/files/16567292/3000WN-1.ZIP)
[3000WN-2.ZIP](https://github.com/user-attachments/files/16567293/3000WN-2.ZIP)
[3000WN-3.ZIP](https://github.com/user-attachments/files/16567294/3000WN-3.ZIP)
[3000WN-4.ZIP](https://github.com/user-attachments/files/16567295/3000WN-4.ZIP)
[3000WN-5.ZIP](https://github.com/user-attachments/files/16567296/3000WN-5.ZIP)

**Reply by stsp** [reply]
Can you tell what is this, before I
d/l 7 archives?

**Reply by jschwartzenberg** [reply]
It's a TCP/IP stack for DOS and Windows by Walker Richer & Quinn Inc.

From the first Windows archive:
```
Walker, Richer & Quinn's 3000 Connection/WIN
TCP/IP stack which is winsock compliant. 1/5
```

From the first DOS archive:
```
Walker, Richer & Quinn's 3000 Connection/DOS
TCP/IP stack which is winsock compliant.
6 disks
(That's right... winsock compatiable from 
 a DOS stack!)
```

So it seems they wanted to combine connectivity across both DOS and Windows, but also made the DOS part winsock (effectively close to BSD sockets?) complaint.

**Reply by stsp** [reply]
https://jeffpar.github.io/kbarchive/kb/080/Q80001/
```
The Reflection Network Series (except Telnet Connection) use the multiprotocol
Network Driver Interface Specification (NDIS) to allow simultaneous connection
to hosts using different protocols. They normally require the following files
loaded into the CONFIG.SYS file:

  device=c:\wrqnet\protman.dos
  device=c:\wrqnet\macwd.dos
```

We don't have NDIS. If we did, we could
as well use native win3.11's winsock.
So unless there are sources of this stack,
its completely useless.

**Reply by jschwartzenberg** [reply]
Yeah, I already thought so. I thought it might be interesting to see another approach that aims for a similar solution on both the DOS and Windows end.

Another large stack plus software set is here: https://winworldpc.com/product/netmanage-chameleon/4x
But judging the screenshot of the older version, it seems to similarly rely on NDIS. I will try to look more into it.

**Comment by stsp** [comment]
Looking into the sock.vxd sources:
https://www.richdawe.be/archive/dossock/csock/csock-20000904.zip
It uses `BlockOnID()` VMM call to stall
the thread and wait for timeouts.
I wonder how can that be done under
dosemu2, where there is no VMM and
the vxds are running inside dosemu2
itself...

@stefand do you know how this
`BlockOnID()` and alike calls can be
implemented? I suppose maybe there
is some int2f API to windows's user-space
scheduler?

**Reply by stsp** [reply]
Ok, I figured in win31 you can suspend
the process only by doing a message
loop - in that case other processes can
work. So winsock for win31 simply never
calls any blocking functions, managing
timeouts by starting the polling timers.
And as we don't have any win95 support,
the TCP interface therefore doesn't need
to care about any windows-friendly
timeout handling.

**Comment by stefand** [comment]
hmm ya I have no idea about these win31/win95 kernel details. Wine doesn't implement any of that.

In terms of who handles the tasks, I would guess most of what the VMM does is either handled by the Linux kernel or dosemu. So in a way there is a VMM. I guess the wait-on-whatever becomes more relevant if you ever implement dos boxes for enhanced mode.

Though are you trying to load that sock.vxd file (whatever it does) or try to implement the API it offers in dosemu so the userspace components can make use of it?

**Reply by stsp** [reply]
> hmm ya I have no idea about these win31/win95 kernel details. Wine doesn't implement any of that.

Some reading revealed there is simply
no such thing. win31 doesn't have a
preemptive scheduler, so the only way
to block, seems to be by running the
even loop. This is also confirmed by
studying the twinsock sources, which
uses polling timers to handle timeouts.
My confusion was raised by the fact that
the aforementioned sock.vxd sources
are for win95. It handles timeouts via
VMM, but under win31 this isn't going
to be used, as calling any blocking VXD
funcs in win31 is strictly forbidden.

> I guess the wait-on-whatever becomes more relevant if you ever implement dos boxes for enhanced mode.

From what I've read, no, completely
impossible in win31. If you block, you
execute event loop by hands, or all
other processes hang. win95 is definitely
more advanced, which is why we don't
run it. :)

> Though are you trying to load that sock.vxd file (whatever it does) or try to implement the API it offers in dosemu
> so the userspace components can make use of it?

dosemu2 already has quite a few vxds
stolen from wine. So yes, sock.vxd is
supposed to be provided by dosemu2
itself. The sources are needed only as
a reference.
Someone may think the tcpabi thing
was a false-start, but I don't think so.
tcpabi spec is good for real-mode
networking apps, but my initial idea
of building a prot-mode networking
around that, was completely crazy.
We definitely need separate real-mode
and prot-mode APIs.

**Comment by stsp** [comment]
> I guess the wait-on-whatever becomes more relevant if you ever implement dos boxes for enhanced mode.

Do you know how to implement that?
Where are the needed vxds documented?
It would be cool to have DOS box in enh
mode.

**Reply by stefand** [reply]
No, I don't really know. I only built a mental model that enhanced mode win3.1 is closer to win95 than DOS from a pile of superficial online articles like

https://devblogs.microsoft.com/oldnewthing/20100517-00/?p=14013
https://www.os2museum.com/wp/vds-borne-out-of-necessity
https://www.os2museum.com/wp/windows-3-x-vddvga

and a few others - I am sure you came across all of that in the past. Somewhere there was also something called "screen grabber" or similar - it sounded to me like a redirection of vga text output and graphics output into a memory buffer so that it could be displayed in a DOS window on the windows desktop.  It would also need something to send keyboard input into the virtual machine. But I have no idea what the actual API for all of those looks like. Maybe Win32's CreateProcess provides a hint to the VM spawn function but I have no clue how the screen and keyboard redirection would work.

Once upon a time I also played with a win16 version of visual c++. It says it needs enhanced mode and tries to install a vxd. Since that vxd isn't working visual studio can't invoke the compiler, although the dos cl.exe works. My guess is that this driver provides some sort of stdout/stderr capture.

**Reply by stsp** [reply]
Yeah, I remember that vga.3gr
driver - we probably don't need
that, running DOS prompt full-screen
is already not so bad.  I suppose
there must be some API for switching
VMs, that is to allow to switch the DPMI
clients in and out. DPMI spec talks about
VMs quite a lot, but no such API is
documented.
I guess it is somewhere in shell.vxd.
Of course our DPMI server allows the
client switching, which is why you can
switch between DPMI DOS progs even
in the std mode DOS prompt of win31.
But adding the enh mode-compatible
VM support would be good.

**Reply by stefand** [reply]
Do you know what the immediate stopper is? I know that when you run a DOS pif file (e.g. the "MS-DOS Prompt" one that just points to command.com I think) it says "Incorrect System Version". Do you know which VMM call it is trying to call and getting an unexpected result?

**Reply by stsp** [reply]
ERROR: DPMI: ERROR: Unsupported VxD 0x445
ERROR: DPMI: ERROR: Unsupported VxD 0x446
ERROR: vxd SHELL: unknown/not implemented parameters:
AX 0000, BX e160, CX 0000, DX 000a, SI 00d6, DI 1e3c, DS 0f8f, ES 0f8f

**Reply by stefand** [reply]
Hmm this nerd-sniped me for a while, but my google-foo didn't find anything of value. I was hoping for old headers that define a name for service 0xA but no luck.

Just a general hint, although you are probably already aware of this: The winevdm project (https://github.com/otya128/winevdm) people might have more knowledge about win16-isms than upstream Wine, and certainly more than I have. That particular shell.vxd call just barfs too though. They probably don't need this thing for running third party Win16 applications.

**Reply by stsp** [reply]
I am not sure if winevdm has anything
to do with win31's DOS box, as that might
be the entirely different technique.
I wonder why MS keeps leaking the
sources of newer windoses, but it
never ever leaked the sources of an
old ones! They might be kept very
secretly there. :)

**Reply by stsp** [reply]
If those 445/446 are external vxds,
it might be possible to figure them
out under qemu, by looking what
file gets loaded by vxdldr...

**Comment by jschwartzenberg** [comment]
Talking about chat clients I know there were prior connectivity issues, but it turns out Session has some security issues as well: https://soatok.blog/2025/01/14/dont-use-session-signal-fork/ 

Did anybody still look at DOSEMU-compatible chat clients in the meantime? :)

**Reply by stsp** [reply]
Session is blocked in Russia long ago.
In dosemu you can likely use IRC under windows.

**Reply by jschwartzenberg** [reply]
> Session is blocked in Russia long ago.

Do you know whether Jami still works?


> In dosemu you can likely use IRC under windows.

I guess only with servers that do not have tight encryption requirements though. Should we establish an IRC channel?

**Reply by stsp** [reply]
Most likely it still works.

**Comment by stsp** [comment]
There is some progress on this
front. I have mostly finished writing
the bsd-compatible sockets lib
for dosemu2: https://github.com/stsp/libcsock
Although right now I can't get
ow debugger to work, so I can't
fix bugs in the code. :)

Also I have fixed the most outstanding
serial port bugs, so I think the
experiment with twinsock should
be retried. IIRC @jschwartzenberg 
wasn't able to get it working last
time. But maybe now it will?

Never thought this TCP stuff is
gonna take that much... First TCP
commit 77cffb3ea is a year old
now.

**Reply by jschwartzenberg** [reply]
> Also I have fixed the most outstanding
> serial port bugs, so I think the
> experiment with twinsock should
> be retried. IIRC @jschwartzenberg
> wasn't able to get it working last
> time. But maybe now it will?

In which sense working? In the screenshot you can see twinsock is working. I did not manage to build its Windows sources myself though. (The unix part is easy.)

**Reply by stsp** [reply]
I only vaguely remember you
did some other experiment and
said twinsock no longer works
but utwinsock does.

**Reply by jschwartzenberg** [reply]
I think they actually ran all at some point again. Its terminal doesn't show anything so that's a bit confusing, but if you blindly start `tshost` everything just works.

**Reply by stsp** [reply]
What serial port mode did
you use on dosemu side?
Exec, pts, or just created a
pty link with socat? Or something
else?

**Reply by jschwartzenberg** [reply]
I used `exec` and set it to `bash`. Then from that shell, I ran `tshost`.

**Reply by stsp** [reply]
Ok, quite strange that `exec`
worked, esp w/o `-no-priv-sep`.
But who knows. :)

**Reply by jschwartzenberg** [reply]
It's been a while. Possibly I used `-no-priv-sep` indeed. I remember getting a shell as a `dosemu2` user at some point and finding a way around this. I think with `-no-priv-sep` got a normal shell as my own user.

**Reply by stsp** [reply]
Please report where ever you
need `-no-priv-sep` as this should
be absolutely never needed, in
an ideal world. But I can't locate
all the subsystems that need to
be ported, on my own.

**Reply by jschwartzenberg** [reply]
I guess for `exec` the proper way would be to make sure that `dosemu2` user could exec the `tshost` binary from twinsock? It would be a matter of proper permissions I'd say.

**Reply by stsp** [reply]
It should just run as your initial
user now, so no actions should
be required at all.
At the worst-case scenario only
some mods to dosemu config
file are needed (like $_unix_exec),
but playing with permissions
should normally not be needed.

**Comment by stsp** [comment]
> Although right now I can't get
> ow debugger to work, so I can't
> fix bugs in the code. :)

Making a progress on this front.
It turned out, the linux port of wd
is broken. So I installed wine to
try wdw - a windows version.
And it works!
Then I connected wine with dosemu
via the PTS link, so that they
have the connected COM ports
between each other.
Then under dosemu I started
`serserv /tr=rsi` for dos4gw-based
debugging, but that was unstable.
Then I tried `serserv /tr=cw` for
causeway-based debugging, and
that works! I can now debug DOS
progs with a GUI debugger started
under wine, and on a source level!

So OW actually offers quite a lot.
Be it GPL, I probably didn't need
to write dj64dev. :)

**Reply by andrewbird** [reply]
Sounds good, are you using the v1.9 openwatcom or the v2.0? If it's the v2.0, then perhaps @jmalak can help with that?

**Reply by stsp** [reply]
He already does, in a different
ticket. :) This ticket is too noisy
for any help.

Another one is this:
https://github.com/open-watcom/open-watcom-v2/issues/1473

**Reply by andrewbird** [reply]
I'm glad you have it in hand!

**Comment by stsp** [comment]
https://github.com/open-watcom/open-watcom-v2/wiki/Remote-debugging-under-Linux-(Ubuntu)
I wrote the instructions on
how to debug with WDW and
dosemu. In case someone is
interested.
